src/sagemaker_sklearn_extension/preprocessing/encoders.py

```diff
@@ -532,10 +557,34 @@ def fit(self, X, y=None):
         self
 
         """
-        # sklearn.preprocessing._BaseEncoder uses _categories due to deprecations in other classes
-        # can be removed once deprecations are removed
-        self._categories = self.categories
-        self._fit(X, handle_unknown="unknown")
+        if isinstance(self.max_categories, int) or self.threshold != 1:
+
+            assert self.max_categories >= 1
+            X_columns, n_samples, n_features = self._check_X(X)
+
+            if self.threshold == "auto":
+                threshold = max(10, n_samples / 1000)
+            elif self.threshold >= 1:
+                threshold = self.threshold
+            else:
+                threshold = ceil(self.threshold * n_samples)
+
+            for i in range(n_features):
+                dtype = X_columns[i].dtype
+                items, counts = np.unique(X_columns[i].astype(str), return_counts=True)
+                categories_to_encode = items[counts >= threshold]
+                if categories_to_encode.size == 0:
+                    warnings.warn(
+                        "feature at index {} does not have any categories appearing more than {} {}".format(
+                            i, threshold, "time" if threshold == 1 else "times"
+                        )
+                    )
+                if len(categories_to_encode) >= self.max_categories:
+                    most_freq_idxs = np.argsort(counts)[len(counts) - self.max_categories + 1 :]
+                    categories_to_encode = items[most_freq_idxs]
+                self.categories.append(np.sort(categories_to_encode.astype(dtype)))
```

Comment:
How do we break ties? i.e. if max_categories is set to 1 but two categorical columns appear equally as much

Reply:
`np.argsort` sorts from min to max and when there is a tie, puts the lower index first. In this code, the index will be the higher index because we take from the back of the array. The index corresponds with the category in the output of `np.unique` which is sorted from min to max. Therefore, if there is a tie between `x` and `y`, two categories of the same frequency, category `y` would be included as `y = max(x, y)`